syntax = "proto3";

package kmeans;

// Service defination for the Mapper
service MapperService{
    rpc perform_map (MapperParameters) returns (MapperResponse);
    rpc send_data_to_reducer (ReducerParameters) returns (ReducerResponse);
}

// Service defination for the Reducer
service ReducerService{
    rpc invoke_reducer (InvokeParameters) returns (InvokeResponse);
}

// Message Definations

// Parameters sent from the Master to Mapper
message MapperParameters {
    string mapper_id = 1;
    int32 start_index = 2;
    int32 end_index = 3;
    repeated Centroid centroids = 4;
    int32 num_reducers = 5;
}  

// Represents a centroid with an ID and coordinates (x, y)
message Centroid {
    string id = 1;
    float x = 2;
    float y = 3;
}

// Response from Mapper to Master
message MapperResponse {
    string status = 1;
}

// Parameters sent from Reducer to Mapper
message ReducerParameters {
    string data_request = 1;
    string reducer_id = 2;
}

// Response from Mapper to Reducer
message ReducerResponse {
    repeated DataPoint data_points = 1;
}

message DataPoint {
  string centroid_id = 1;
  string x = 2;
  string y = 3;
}

// Parameters sent from Master to Reducer
message InvokeParameters {
    int32 num_mappers = 1;
    int32 num_centroids = 2;
}

// Response from Reducer to Master
message InvokeResponse {
    map<string, Points> data = 1;
}


message Point {
  string x = 1;
  string y = 2;
}

message Points {
  repeated Point points = 1;
}
